**If you use SNPTB, please add the following information to your manuscript:**

_"Material and Methods" (please rephrase in your own words):_

SNPs were identified from the whole genome sequencing data using the bioinformatics pipeline SNPTB [1]. In brief, Trimmomatic (version 0.36) was used to do the quality control analysis and the filtered reads were mapped to the H37Rv reference genome (NCBI Accession AL123456.3) using Bowtie 2 (version 2.2.6) [2,3]. High-confidence SNPs were identified using SAMtools (version 1.2) and BCFtools (version 1.2) [4,5].



_Please cite ALL of the following manuscripts to give due credit to the open-source software that SNPTB uses:_

REFERENCES

1. Aditi Gupta (2017). SNPTB: nucleotide variant identification and annotation in Mycobacterium tuberculosis genomes. bioRxiv, doi:10.1101/227066.
2. Anthony M. Bolger, Marc Lohse, Bjoern Usadel; Trimmomatic: a flexible trimmer for Illumina sequence data, Bioinformatics, Volume 30, Issue 15, 1 August 2014, Pages 2114â€“2120
3. Langmead B, Salzberg S. Fast gapped-read alignment with Bowtie 2. Nature Methods. 2012, 9:357-359
4. Li H, A statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data, Bioinformatics (2011) 27(21) 2987-93.
5. BCFtools/RoH: a hidden Markov model approach for detecting autozygosity from next-generation sequencing data. Narasimhan V, Danecek P, Scally A, Xue Y, Tyler-Smith C and Durbin R Bioinformatics (Oxford, England) 2016;32;11;1749-51
